test/e2e/suite/secrettemplate/secrettemplate.go

```diff
@@ -0,0 +1,453 @@
+		Expect(err).NotTo(HaveOccurred())
+
+		if secret.Annotations == nil {
+			secret.Annotations = make(map[string]string)
+		}
+		if secret.Labels == nil {
+			secret.Labels = make(map[string]string)
+		}
+		secret.Annotations["an-annotation"] = "bar"
+		secret.Annotations["another-annotation"] = "def"
+		secret.Labels["abc"] = "123"
+		secret.Labels["foo"] = "bar"
+
+		secret, err = f.KubeClientSet.CoreV1().Secrets(f.Namespace.Name).Update(context.Background(), secret, metav1.UpdateOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		secret, err = f.KubeClientSet.CoreV1().Secrets(f.Namespace.Name).Get(context.Background(), secretName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(secret.Annotations).To(HaveKeyWithValue("an-annotation", "bar"))
+		Expect(secret.Annotations).To(HaveKeyWithValue("another-annotation", "def"))
+		Expect(secret.Labels).To(HaveKeyWithValue("abc", "123"))
+		Expect(secret.Labels).To(HaveKeyWithValue("foo", "bar"))
+
+		secret, err = f.KubeClientSet.CoreV1().Secrets(f.Namespace.Name).Apply(context.Background(),
+			applycorev1.Secret(secret.Name, secret.Namespace).
+				WithAnnotations(secret.Annotations).
+				WithLabels(secret.Labels),
+			metav1.ApplyOptions{FieldManager: "e2e-test-client"})
+
+		By("expect those Annotations and Labels to be present on the Secret")
+		secret, err = f.KubeClientSet.CoreV1().Secrets(f.Namespace.Name).Get(context.Background(), secretName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(secret.Annotations).To(HaveKeyWithValue("an-annotation", "bar"))
+		Expect(secret.Annotations).To(HaveKeyWithValue("another-annotation", "def"))
+		Expect(secret.Labels).To(HaveKeyWithValue("abc", "123"))
+		Expect(secret.Labels).To(HaveKeyWithValue("foo", "bar"))
+
+		By("add those Annotations and Labels to the SecretTemplate of the Certificate")
+		crt.Spec.SecretTemplate = &cmapi.CertificateSecretTemplate{
+			Annotations: map[string]string{"an-annotation": "bar", "another-annotation": "def"},
+			Labels:      map[string]string{"abc": "123", "foo": "bar"},
+		}
+
+		crt, err = f.CertManagerClientSet.CertmanagerV1().Certificates(f.Namespace.Name).Update(context.Background(), crt, metav1.UpdateOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		By("waiting for those Annotation and Labels on the Secret to contain managed fields from cert-manager")
+
+		Eventually(func() bool {
+			secret, err = f.KubeClientSet.CoreV1().Secrets(f.Namespace.Name).Get(context.Background(), secretName, metav1.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			var managedLabels, managedAnnotations []string
+			for _, managedField := range secret.ManagedFields {
+				// The manager of the controller is currently "controller" which is
+				// based on the user agent of the controller binary. This manager name
+				// would change if the User Agent is changed.
+				if managedField.Manager != "controller" || managedField.FieldsV1 == nil {
+					continue
+				}
```

Comment:
I am confused with the string `"controller"`, I thought it would be more precise, such as `secret-manager`. Shouldn't we have one client per controller so that the apiserver can deal with inter-controller conflicts?

Right now, the server-side apply operation coming from the "secret template controller" looks like this:

![Screenshot from 2022-01-15 18-46-02](https://user-images.githubusercontent.com/2195781/149632231-35e00663-a017-4a66-8c91-3e06cfc8e154.png)

From:

> Managers identify distinct workflows that are modifying the object (especially useful on conflicts!), and can be specified through the `fieldManager` query parameter as part of a modifying request. It is required for the apply endpoint, though kubectl will default it to kubectl. For other updates, its default is computed from the user-agent.

So I guess the user agent can stay the same `controller/.....`, but shouldn't we change the `fieldManager` query param (encircled in red in the above screenshot) to something like `secret-template-controller`?

Reply:
The end goal here is that we want all controllers in cert-manager to be using the Apply operation, and as you've pointed out, we want all of these controllers to be using distinct field managers.

Unless specified, the field manager string will be derived from the Kubernetes client's user-agent. This is quite nice behaviour since we get a stronger audit/metric trail of who's doing what (since the same string will be used for both Apply, as well as other operations), so this is how I think the field manager should be configured.

I wanted to keep this PR scoped to just adding Apply to the SecretsManager here, but intend to make follow-up PRs for both changing API calls to Apply, as well as setting the user agent individually for each controller.

Here's some more discussion on the topic: 
https://github.com/jetstack/cert-manager/issues/4552
https://kubernetes.slack.com/archives/CDEQJ0Q8M/p1642176983252100

Reply:
> Unless specified, the field manager string will be derived from the Kubernetes client's user-agent. This is quite nice behaviour since we get a stronger audit/metric trail of who's doing what (since the same string will be used for both Apply, as well as other operations), so this is how I think the field manager should be configured.

That makes sense.

- Since the secret template controller is the only controller to make apply calls, this is not a problem to set the `fieldManager` to `controller`.
- My understanding is that we will be later on using one user-agent per controller. For non-apply calls, the user-agent will be use by the rest client, and apply-specific code will derive the fieldManager from the controller-specific user-agent. This will allow us to harmonize the user agents across non-apply and apply calls, which is a good thing.

Thanks for the precisions.

Reply:
Yep, that's exactly it :+1: